### ISSUE TITLE
Add a file-wide keep pragma

### ISSUE THREAD

**Issue by Flamefire** [include-what-you-use/include-what-you-use#267]
I have some include files, that define generic stuff or something for the internal build system in some cases. Once example is the `DEBUG_NEW` define for MSVC (basically redefines `new` to include information on an allocation) I put that in a header file and enforce (via code guide) that it must be included in every cpp file.
However iwyu suggests removing this (at least on linux) probably because the pure windows stuff does not get through the preprocessor. I could add a `keep` pragma to every inclusion, but this would be quite some useless work.
So I suggest adding a macro to the header file itself, that tells iwyu that it should not suggest removing it even if it seems to be possible.

**Comment by kimgr** [comment]
It sounds like an `IWYU pragma: keep-always` added to the header would be helpful here.

I can see how that could also be useful e.g. for `config.h` headers, precompiled headers and other constructs that we want to keep whether they provide used symbols or not.

Or maybe a command-line argument, e.g. `--keep_always=xyz.h`?

I'm just throwing out ideas here, I think we have enough pragmas and command-line args already, but it seems like a useful feature.

**Comment by Flamefire** [comment]
I'd rather like a pragma so you can't forget it e.g. in team projects

**Comment by vsapsai** [comment]
Is this file prefix header in Linux build? I am thinking about options to detect such cases automatically from compilation flags.

**Comment by Flamefire** [comment]
It is a cross-platform project and in this case it is a linux build yes because it is easier to seup for iwyu (is it even possible on windows?). One of the headers is empty in linux builds (via ifdefs)

**Comment by vsapsai** [comment]
Are options
```
   --pch_in_code: mark the first include in a translation unit as a
        precompiled header.  Use --pch_in_code to prevent IWYU from
        removing necessary PCH includes.  Though Clang forces PCHs
        to be listed as prefix headers, the PCH-in-code pattern can
        be used with GCC and is standard practice on MSVC
        (e.g. stdafx.h).
   --prefix_header_includes=<value>: tells iwyu what to do with
        in-source includes and forward declarations involving
        prefix headers.  Prefix header is a file included via
        command-line option -include.  If prefix header makes
        include or forward declaration obsolete, presence of such
        include can be controlled with the following values
          add:    new lines are added
          keep:   new lines aren't added, existing are kept intact
          remove: new lines aren't added, existing are removed
        Default value is 'add'.
```

applicable in your case? It sounds like they are relevant to your setup but don't know if they are useful for you.

**Comment by Flamefire** [comment]
Nope, they are not helpful. The project consists of subprojects and not all of them have that special header. Still trying to compile them separately is not easily possible.

**Comment by RalphS69** [comment]
Afaiu the option `--keep` should do the trick (available at least in version 13.0.0, which is the current head for clang version 13 - it might not have been available six years ago):

```
    --keep=<glob>: tells iwyu to always keep these includes.
         This flag may be specified multiple times to specify
         multiple glob patterns.
```

However, I don't get this option to do what it imho should do. With
```
./include-what-you-use -Xiwyu --keep=options.h my_module.c
```
I get exactly the same as without the `--keep` option, i.e. the `#include "options.h"` in `my_module.h`(being included from `my_module.c` gets removed. Any help on this matter would be appreciated.

BTW: An alternative would be, if the option `--safe_header` of `fix_includes.py` would be configurable to just keep dedicated includes ... then I would not have to run through the whole process twice (I just want to keep `#include options.h` in my header files, but not in my c files).

**Comment by kimgr** [comment]
@RalphS69 Looks like it should just work. What happens if you up the verbosity level, to e.g. `-Xiwyu --verbose=6`? Any clues?
